### ISSUE TITLE
Nicolas Chagnet's Page

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#11988]
Nicolas Chagnet's Page<br>
<br>
https://ift.tt/SfKmZLB<br>
<br>
<br>
<br>

<div><div>
<section><h4 id="interlude-preprocessing-and-feature-engineering">Interlude: preprocessing and feature engineering</h4>
<p>The module <code>preprocessing.py</code> contains two important classes to handle data prior to doing any training or prediction. The class <code>LinearlyInterpolateTS</code> helps with missing data in the downloaded timeseries. The second class, <code>ExogBuilder</code>, creates exogeneous features based on the index of the timeseries. I determined which calendar features to consider when exploring the data and settled with:</p>
<ul>
<li>a <code>holidays</code> boolean column, built using the <a href="https://pypi.org/project/holidays/">holidays</a> Python package, which indicates which days are French national holidays,</li>
<li>an <code>is_weekend</code> boolean column indicating whether the day of the week in Saturday or Sunday,</li>
<li>multiple columns <span><span><math>TiT_i</math></span></span> where <span><span><math>TT</math></span></span> indicates a seasonality (period) and <span><span><math>ii</math></span></span> is the order of the encoding.</li>
</ul>
<p>This last set of features is crucial to really capture all of the seasonality for this kind of data. During the exploratory phase, I found various seasonalities in the data: energy demand generally depends on the seasonal temperature variations, the day/night cycle, and the work/weekend cycle. There are multiple ways to determine the various frequencies present in the data, like looking at autocorrelation plots or superimposing data grouped by various periodicities (for example, yearly data plotted by month, etc…). My favorite method is to use a simple Fourier analysis. Here is an example of how you can do this using <a href="https://docs.scipy.org/doc/scipy/reference/signal.html">scipy.signal</a>, with a plot of the resulting spectrum:</p>
<div><pre tabindex="0"><code><span><span>from</span><span> scipy.signal </span><span>import</span><span> periodogram</span></span>
<span></span>
<span><span>frequencies, spectrum </span><span>=</span><span> periodogram(</span></span>
<span><span>    y, </span><span># Timeseries</span></span>
<span><span>    window</span><span>=</span><span>"boxcar"</span><span>,</span></span>
<span><span>    scaling</span><span>=</span><span>"spectrum"</span><span>,</span></span>
<span><span>)</span></span></code></pre></div>
<p><img src="https://nchagnet.pages.dev/assets/blog/energy-forecast-project/periodogram.png" alt="Periodogram"></p>
<p>The peaks on this graph indicate the frequencies which are the most important for the data: there is a strong annual signal and slightly weaker bi-annual seasonality which match the pattern between summer and winter demand. There are also well-defined weekly and daily seasonalities, and even a twice-daily peak which is usually a sign of peak/off-peak times. Finally there is some signal in the monthly range which is less sharp, showing more variation than the others.</p>
<p>There are a few ways to account for these patterns in the exogeneous features, and a good summary of these is present in the <a href="https://skforecast.org/0.14.0/faq/cyclical-features-time-series">Skforecast documentation</a>. The basic idea with cyclical encoding is that, if you have a feature <code>X[col]</code> which you know is periodic, then you want to somehow encode that the last value is close to the first value of your period. For example, with hours, 01:00 is as close to 23:00 as it is to 03:00. Or Sunday (7) being close to both Monday (1) and Saturday (6). Ordinal encoding just doesn’t reflect that.</p>
<p>I have often used Fourier decompositions which can be neatly done through the <a href="https://www.statsmodels.org/stable/index.html">statsmodels</a> library. However, I have found better performance with the radial basis functions decomposition, which you can do with the preprocessing module of the <a href="https://koaning.github.io/scikit-lego/">sklego</a> library.</p>
<p>With Fourier decomposition, you just replace your column by two new features: a cosine and a sine with the same period. Then your model learns from them to recognize the seasonality (daily, weekly, etc…).</p>
<p>Radial basis functions work the same way, except that instead of adding two new trigonometric features, you can add <span><span><math>NN</math></span></span> shifted periodic Gaussian features. In practice, this encoding seems to yield better results than Fourier modes. The number of functions added is a parameter you can choose: in this project I tended to use the periodicity itself (7 in a week, 12 in a month) if it’s small enough, or a sensible divider: 12 for the hours or 12 for the days in the year (with each function encodes the beginning of a month).</p>
<p>There’s another set of useful features that can be included to improve model accuracy which I mentioned before:</p>
<ul>
<li>lags, which are shifted versions of the timeseries itself,</li>
<li>window features, which are target features calculated on a rolling window over the data.</li>
</ul>
<p>Since both of these are based on target values, the Skforecast module handles this through the forecaster object directly. This is what the lags parameter is for in the <code>ForecasterRecursive</code> object, and there is a similar one for window features.</p>
</section><section><h4 id="back-to-model-training">Back to model training</h4>
<p>With the exogeneous features set, you still need to choose a model. For this type of problem, boosting-based models tend to work quite well, so I made the wrapper compatible with both XGBoost and LGBM regressors. Which one to choose is a matter of testing which one works better, and from my experiments I ended up favoring LGBM as it seemed to be overfitting less than XGBoost.<sup><a href="https://nchagnet.pages.dev/blog/energy-forecast-project/#user-content-fn-overfit" id="user-content-fnref-overfit">5</a></sup> In both cases, I included L1 and L2 regularization parameters, which help to improve the generalization of the model.</p>
<p>Finally, it’s time to tune the model’s hyperparameters on a cross-validation dev set before setting it in production. The tuning uses Bayesian search on an <a href="https://macalusojeff.github.io/post/HyperparameterTuningLGBM/">appropriate range</a> of parameters, as well as the number of lags to include. With Skforecast, you can do hyperparameter tuning using causal cross-validation folds, <em>i.e.</em>, separate data into folds of training and dev datasets where all the training data is before the dev data. Here is a small example of how to create these folds:</p>
<div><pre tabindex="0"><code><span><span>from</span><span> skforecast.model_selection </span><span>import</span><span> TimeSeriesFold</span></span>
<span></span>
<span><span>cv </span><span>=</span><span> TimeSeriesFold(</span></span>
<span><span>    steps</span><span>=</span><span>24</span><span>, </span><span># Number of steps to predict</span></span>
<span><span>    refit</span><span>=</span><span>False</span><span>, </span><span># Whether to refit between each fold. If an int is given refit intermittently</span></span>
<span><span>    initial_train_size</span><span>=</span><span>24</span><span>*</span><span>365</span><span>*</span><span>3</span><span>, </span><span># Train over 3 years of data</span></span>
<span><span>    allow_incomplete_fold</span><span>=</span><span>True</span><span>, </span><span># If the last fold is not complete, still use it</span></span>
<span><span>)</span></span></code></pre></div>
<p>This cross-validation object can then be used when backtesting or when searching for hyperparameters. Here is an example on how to do the latter:</p>
<div><pre tabindex="0"><code><span><span>from</span><span> skforecast.model_selection </span><span>import</span><span> bayesian_search_forecaster</span></span>
<span><span>from</span><span> optuna.trial </span><span>import</span><span> Trial</span></span>
<span></span>
<span><span>def</span><span> search_space_lgbm</span><span>(trial: Trial) -&gt; </span><span>dict</span><span>:</span></span>
<span><span>  """Function describing the search space for hyperparameters of an LGBM model."""</span></span>
<span><span>  search_space </span><span>=</span><span> {</span></span>
<span><span>      "num_leaves"</span><span>: trial.suggest_int(</span><span>"num_leaves"</span><span>, </span><span>8</span><span>, </span><span>256</span><span>),</span></span>
<span><span>      "max_depth"</span><span>: trial.suggest_int(</span><span>"max_depth"</span><span>, </span><span>3</span><span>, </span><span>16</span><span>),</span></span>
<span><span>      "learning_rate"</span><span>: trial.suggest_float(</span><span>"learning_rate"</span><span>, </span><span>0.001</span><span>, </span><span>0.2</span><span>, </span><span>log</span><span>=</span><span>True</span><span>),</span></span>
<span><span>      "n_estimators"</span><span>: trial.suggest_int(</span><span>"n_estimators"</span><span>, </span><span>50</span><span>, </span><span>1000</span><span>, </span><span>log</span><span>=</span><span>True</span><span>),</span></span>
<span><span>      "bagging_fraction"</span><span>: trial.suggest_float(</span><span>"bagging_fraction"</span><span>, </span><span>0.5</span><span>, </span><span>1</span><span>),</span></span>
<span><span>      "feature_fraction"</span><span>: trial.suggest_float(</span><span>"feature_fraction"</span><span>, </span><span>0.5</span><span>, </span><span>1</span><span>),</span></span>
<span><span>      "reg_alpha"</span><span>: trial.suggest_float(</span><span>"reg_alpha"</span><span>, </span><span>0.01</span><span>, </span><span>100</span><span>),</span></span>
<span><span>      "reg_lambda"</span><span>: trial.suggest_float(</span><span>"reg_lambda"</span><span>, </span><span>0.01</span><span>, </span><span>100</span><span>),</span></span>
<span><span>      "lags"</span><span>: trial.suggest_categorical(</span><span>"lags"</span><span>, c.lags_consider),</span></span>
<span><span>  }</span></span>
<span><span>  return</span><span> search_space</span></span>
<span></span>
<span><span># The Bayesian search forecaster uses Optuna</span></span>
<span><span>results, _ </span><span>=</span><span> bayesian_search_forecaster(</span></span>
<span><span>    forecaster</span><span>=</span><span>forecaster, </span><span># The forecaster object</span></span>
<span><span>    y</span><span>=</span><span>y, </span><span># Timeseries data of shape (N,1)</span></span>
<span><span>    cv</span><span>=</span><span>cv, </span><span># The cross-validation folds</span></span>
<span><span>    search_space</span><span>=</span><span>search_space_lgbm, </span><span># Range of parameters to consider</span></span>
<span><span>    metric</span><span>=</span><span>"mean_absolute_error"</span><span>,</span></span>
<span><span>    exog</span><span>=</span><span>X, </span><span># Exogeneous features of shape (N, n_features)</span></span>
<span><span>    return_best</span><span>=</span><span>False</span><span>, </span><span># If True, the model will be retrained on the best parameters found</span></span>
<span><span>    random_state</span><span>=</span><span>c.random_state,</span></span>
<span><span>    verbose</span><span>=</span><span>False</span><span>,</span></span>
<span><span>    n_trials</span><span>=</span><span>10</span><span>, </span><span># Number of trials to find the right parameters</span></span>
<span><span>)</span></span>
<span></span>
<span><span>print</span><span>(</span><span>f</span><span>"Best parameters:</span><span>\n</span><span> {</span><span>results.iloc[</span><span>0</span><span>]</span><span>}</span><span>"</span><span>)</span></span></code></pre></div>
<p>There are various strategies you can choose from in cross-validation: whether you want to refit the regressor between each fold, intermittently, or never, but also whether you want to keep the training size fixed, etc… The preceding example only showed a sample of what’s possible, and I recommend going through <a href="https://skforecast.org/0.14.0/introduction-forecasting/introduction-forecasting#backtesting-forecasting-models">the Skforecast documentation</a> for more details.</p>
<p>My advice is to always choose a strategy that resembles as much as possible the actual prediction scenario. In my case, the program will predict batches of 24 hours of data, with retraining every 7 days, so I chose to work with a fold size of 24, and the model should be refitted every 7 folds. However, to reduce the computational load, I also chose not to refit the model during cross-validation, even though it would be more accurate: sometimes you have to make some compromises for performance!</p>
<p>Once the model is tuned, the <code>ForecasterRecursiveModel</code> class is then fitted on the whole <code>training + dev</code> sets of data before making new predictions. Finally, the model is serialized to a file using the <a href="https://pypi.org/project/dill/">dill</a> library, so it can be easily loaded when it’s time to make inferences. I initially used <a href="https://joblib.readthedocs.io/en/stable/">joblib</a> for this part, but after refactoring the code, it somehow stopped working. I couldn’t find the exact reason why the joblib library suddenly failed, but as usual the dill library worked when joblib wouldn’t. However, this isn’t a complete fix to me, and an improved version of this function would be to save only the regressor instead of dumping the whole <code>ForecasterRecursiveLGBM</code> object.</p>
</section>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via nchagnet.pages.dev https://ift.tt/3e2GFIh<br>
<br>
February 17, 2025 at 11:33AM
